full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,155 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function loadTodos() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) throw new Error(`fetch error ${response.status}`);
+        const todos = (await response.json()) as Todo[];
+        setTodos(todos);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    loadTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos', {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(newTodo),
+      });
+      if (!response.ok) {
+        throw new Error(`fetch err, status: ${response.status}`);
+      }
+      const todo = await response.json();
+      setTodos([...todos, todo]);
+    } catch (err) {
+      setError(err);
```

Comment:
This is fine here, but it's probably not the best user experience since it will cause the entire page to disappear and be replaced with the error message. There are a number of ways to display errors in a friendly way, but we don't cover any of them at LFZ. For now, a reasonable error UI is to just log and use `alert()`.